rust/main/hyperlane-metric/src/sealevel/prometheus_sender.rs

```diff
@@ -0,0 +1,160 @@
+use std::{
+    sync::{atomic::Ordering, Arc},
+    time::{Duration, Instant},
+};
+
+use maplit::hashmap;
+use reqwest::{
+    header::{CONTENT_TYPE, RETRY_AFTER},
+    StatusCode,
+};
+use solana_client::{
+    client_error::ClientError,
+    rpc_request::{RpcError, RpcRequest},
+    rpc_sender::{RpcSender, RpcTransportStats},
+};
+use tokio::time::sleep;
+use url::Url;
+
+use crate::prometheus_metric::{
+    JsonRpcClientMetrics, PrometheusJsonRpcClientConfig, PrometheusJsonRpcClientConfigExt,
+};
+use crate::sealevel::http_sender::{rpc_error_object_to_response, HttpSender, RpcErrorObject};
+
+pub struct SealevelRpcSender {
+    pub inner: HttpSender,
+}
+
+pub struct PrometheusSealevelRpcSender {
+    pub inner: HttpSender,
+    pub metrics: JsonRpcClientMetrics,
+    pub config: PrometheusJsonRpcClientConfig,
+}
+
+impl PrometheusSealevelRpcSender {
+    pub fn new(
+        url: Url,
+        metrics: JsonRpcClientMetrics,
+        config: PrometheusJsonRpcClientConfig,
+    ) -> Self {
+        Self {
+            inner: HttpSender::new(url),
+            metrics,
+            config,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl RpcSender for PrometheusSealevelRpcSender {
+    fn get_transport_stats(&self) -> RpcTransportStats {
+        self.inner.stats.read().unwrap().clone()
+    }
+
+    async fn send(
+        &self,
+        request: RpcRequest,
+        params: serde_json::Value,
+    ) -> Result<serde_json::Value, ClientError> {
+        let start = Instant::now();
+        let request_id = self.inner.request_id.fetch_add(1, Ordering::Relaxed);
+        let method = format!("{}", request);
+        let res =
+            send_sealevel_rpc_request(&self.inner.client, self.url(), request_id, request, params)
+                .await;
+
+        let labels = hashmap! {
+            "provider_node" => self.config.node_host(),
+            "chain" => self.config.chain_name(),
+            "method" => &method,
+            "status" => if res.is_ok() { "success" } else { "failure" }
+        };
+        if let Some(counter) = &self.metrics.request_count {
+            counter.with(&labels).inc()
+        }
+        if let Some(counter) = &self.metrics.request_duration_seconds {
+            counter
+                .with(&labels)
+                .inc_by((Instant::now() - start).as_secs_f64())
+        };
```

Comment:
this logic looks like the one for instatiating the ethers metrics provider, maybe it can be refactored

Reply:
Moved into JsonMetrics method :+1: